File: src/studycase/fms/esp32/product_module.cpp
#include "product_module.hpp"

NextSkillBehavior::NextSkillBehavior(ProductAgent* agent)
    : SimpleBehavior(), productAgent(agent), requestSent(false), searchCompleted(false) {
    emaInteraction = new tamaf::ema::EMAInteraction(this);
}

NextSkillBehavior::~NextSkillBehavior() {
    delete emaInteraction;
}

void NextSkillBehavior::Action() {
    if (!productAgent) {
        searchCompleted = true;
        return;
    }

    String skill = productAgent->currentSkill();
    if (skill.length() == 0) {
        Serial.print("[");
        Serial.print(productAgent->GetAgentID().getName());
        Serial.println("] All skills completed.");
        searchCompleted = true;
        return;
    }

    if (!requestSent) {
        Serial.print("[");
        Serial.print(productAgent->GetAgentID().getName());
        Serial.print("] Searching for skill: ");
        Serial.println(skill);

        tamaf::types::ServiceDescription sd(skill.c_str(), "");
        searchAD.addService(sd);
        requestSent = true;
    }

    auto result = emaInteraction->Search(searchAD);
    if (result.has_value()) {
        searchCompleted = true;
        if (result.value().empty()) {
            Serial.print("[");
            Serial.print(productAgent->GetAgentID().getName());
            Serial.println("] No agents found for skill. Retrying in 1s...");
            delay(1000);
            productAgent->AddBehavior(new NextSkillBehavior(productAgent));
        } else {
            tamaf::messaging::ACLMessage cfp(tamaf::enums::Performative::CFP);
            for (const auto& ad : result.value()) {
                cfp.addReceiver(ad.getAgentId());
            }
            cfp.setOntology(ONTOLOGY_NEGOTIATE_NEXT_RESOURCE);
            cfp.setContent(skill.c_str());

            productAgent->AddBehavior(new CNETProposalInitiator(productAgent, cfp));
        }
    }
}

// CNETProposalInitiator
CNETProposalInitiator::CNETProposalInitiator(ProductAgent* agent, const tamaf::messaging::ACLMessage& cfpMsg)
    : FIPAContractNetInitiator(cfpMsg), productAgent(agent) {
    emaInteraction = new tamaf::ema::EMAInteraction(this);
}

CNETProposalInitiator::~CNETProposalInitiator() {
    delete emaInteraction;
}

void CNETProposalInitiator::handleProposals(const std::vector<tamaf::messaging::ACLMessage>& proposals, std::vector<tamaf::messaging::ACLMessage>& responses) {
    if (proposals.empty()) {
        Serial.print("[");
        Serial.print(productAgent->GetAgentID().getName());
        Serial.println("] No proposals received.");
        return;
    }

    // Select the first proposal as best
    const auto& bestProposal = proposals[0];
    tamaf::messaging::ACLMessage accept(tamaf::enums::Performative::ACCEPT_PROPOSAL);
    accept.addReceiver(bestProposal.getSender());
    accept.setConversationId(bestProposal.getConversationId());
    accept.setOntology(bestProposal.getOntology());
    responses.push_back(accept);

    // Reject others
    for (size_t i = 1; i < proposals.size(); ++i) {
        tamaf::messaging::ACLMessage reject(tamaf::enums::Performative::REJECT_PROPOSAL);
        reject.addReceiver(proposals[i].getSender());
        reject.setConversationId(proposals[i].getConversationId());
        reject.setOntology(proposals[i].getOntology());
        responses.push_back(reject);
    }
}

void CNETProposalInitiator::handleInform(const tamaf::messaging::ACLMessage& informMsg) {
    if (!productAgent) return;

    productAgent->setResourceExecuter(informMsg.getSender());
    Location nextLoc = stringToLocation(informMsg.getContent().c_str());
    productAgent->setNextLocation(nextLoc);

    Serial.print("[");
    Serial.print(productAgent->GetAgentID().getName());
    Serial.print("] Proposal accepted. Next location: ");
    Serial.println(locationToString(nextLoc));

    // Search for transport agent
    tamaf::types::AgentDescription searchAD;
    tamaf::types::ServiceDescription sd(DF_SERVICE_NAME_TRANSPORT, DF_SERVICE_TYPE_TRANSPORT);
    searchAD.addService(sd);

    auto result = emaInteraction->Search(searchAD);
    if (!result.has_value() || result.value().empty()) {
        Serial.print("[");
        Serial.print(productAgent->GetAgentID().getName());
        Serial.println("] Transport agent not found!");
        return;
    }

    tamaf::messaging::ACLMessage req(tamaf::enums::Performative::REQUEST);
    req.addReceiver(result.value()[0].getAgentId());
    req.setOntology(ONTOLOGY_REQUEST_TRANSPORT);

    String content = "{\"a\":\"" + locationToString(productAgent->getMyLocation()) + "\",\"b\":\"" + locationToString(nextLoc) + "\"}";
    req.setContent(content.c_str());

    productAgent->AddBehavior(new REQUESTExecuteTransportInitiator(productAgent, req));
}

// REQUESTExecuteTransportInitiator
REQUESTExecuteTransportInitiator::REQUESTExecuteTransportInitiator(ProductAgent* agent, const tamaf::messaging::ACLMessage& request)
    : FIPARequestInitiator(request), productAgent(agent) {}

void REQUESTExecuteTransportInitiator::HandleInform(const tamaf::messaging::ACLMessage& inform) {
    if (!productAgent) return;

    productAgent->setMyLocation(productAgent->getNextLocation());
    Serial.print("[");
    Serial.print(productAgent->GetAgentID().getName());
    Serial.print("] Transport completed. Arrived at ");
    Serial.println(locationToString(productAgent->getMyLocation()));

    tamaf::messaging::ACLMessage req(tamaf::enums::Performative::REQUEST);
    req.addReceiver(productAgent->getResourceExecuter());
    req.setOntology(ONTOLOGY_REQUEST_EXECUTE_SKILL);
    req.setContent(productAgent->currentSkill().c_str());

    productAgent->AddBehavior(new REQUESTExecuteSkillInitiator(productAgent, req));
}

void REQUESTExecuteTransportInitiator::HandleRefuse(const tamaf::messaging::ACLMessage& refuse) {
    if (!productAgent) return;
    Serial.print("[");
    Serial.print(productAgent->GetAgentID().getName());
    Serial.println("] Transport refused. Retrying...");
    productAgent->AddBehavior(new NextSkillBehavior(productAgent));
}

// REQUESTExecuteSkillInitiator
REQUESTExecuteSkillInitiator::REQUESTExecuteSkillInitiator(ProductAgent* agent, const tamaf::messaging::ACLMessage& request)
    : FIPARequestInitiator(request), productAgent(agent) {}

void REQUESTExecuteSkillInitiator::HandleInform(const tamaf::messaging::ACLMessage& inform) {
    if (!productAgent) return;

    Serial.print("[");
    Serial.print(productAgent->GetAgentID().getName());
    Serial.print("] Skill ");
    Serial.print(productAgent->currentSkill());
    Serial.println(" completed.");

    if (productAgent->hasNextSkill()) {
        productAgent->setNextSkill();
        productAgent->AddBehavior(new NextSkillBehavior(productAgent));
    } else {
        Serial.print("[");
        Serial.print(productAgent->GetAgentID().getName());
        Serial.println("] All skills completed. Returning to location A.");

        tamaf::messaging::ACLMessage req(tamaf::enums::Performative::REQUEST);
        req.setOntology(ONTOLOGY_REQUEST_TRANSPORT);
        String content = "{\"a\":\"" + locationToString(productAgent->getMyLocation()) + "\",\"b\":\"A\"}";
        req.setContent(content.c_str());

        productAgent->AddBehavior(new TransportBackBehavior(productAgent, req));
    }
}

void REQUESTExecuteSkillInitiator::HandleRefuse(const tamaf::messaging::ACLMessage& refuse) {
    if (!productAgent) return;
    Serial.print("[");
    Serial.print(productAgent->GetAgentID().getName());
    Serial.println("] Skill execution refused. Retrying...");
    productAgent->AddBehavior(new NextSkillBehavior(productAgent));
}

// TransportBackBehavior
TransportBackBehavior::TransportBackBehavior(ProductAgent* agent, const tamaf::messaging::ACLMessage& request)
    : FIPARequestInitiator(request), productAgent(agent) {}

void TransportBackBehavior::HandleInform(const tamaf::messaging::ACLMessage& inform) {
    if (!productAgent) return;
    Serial.print("[");
    Serial.print(productAgent->GetAgentID().getName());
    Serial.println("] Back at A. Work completed successfully!");
}

void TransportBackBehavior::HandleFailure(const tamaf::messaging::ACLMessage& failure) {
    if (!productAgent) return;
    Serial.print("[");
    Serial.print(productAgent->GetAgentID().getName());
    Serial.println("] Failed to return to A.");
}

// ProductAgent
ProductAgent::ProductAgent(const String& name, const std::vector<String>& process, Location loc, const String& ip)
    : Agent(formatAgentName(name).c_str(), tamaf::types::Address(ip, 4001)),
      myProcess(process), myLocation(loc), nextLocation(loc), currentSkillIndex(0),
      resourceExecuter("", tamaf::types::Address("0.0.0.0", 0)) {}

void ProductAgent::Setup(tamaf::ema::EMAInteraction* emaInteraction) {
    AddBehavior(new NextSkillBehavior(this));
}

String ProductAgent::currentSkill() const {
    if (currentSkillIndex < myProcess.size()) {
        return myProcess[currentSkillIndex];
    }
    return "";
}

bool ProductAgent::hasNextSkill() const {
    return (currentSkillIndex + 1) < myProcess.size();
}

void ProductAgent::setNextSkill() {
    currentSkillIndex++;
}

File: src/studycase/fms/python/product_module.py
import json
import time
from tamaf import (
    Agent, 
    OneShotBehavior, 
    FIPAContractNetInitiator, 
    FIPARequestInitiator, 
    Performative, 
    AgentDescription, 
    ServiceDescription,
    ACLMessage
)
from constants import (
    DF_SERVICE_TYPE_SKILL,
    DF_SERVICE_TYPE_TRANSPORT,
    DF_SERVICE_NAME_TRANSPORT,
    ONTOLOGY_NEGOTIATE_NEXT_RESOURCE,
    ONTOLOGY_REQUEST_TRANSPORT,
    ONTOLOGY_REQUEST_EXECUTE_SKILL,
    get_host_tag
)

class ProductAgent(Agent):
    def __init__(self, name, process, location):
        """
        :param name: Agent name
        :param process: List of skill names (strings)
        :param location: Initial location (string key from Locations)
        """
        host_tag = get_host_tag()
        if not name.endswith(f"_{host_tag}"):
            name = f"{name}_{host_tag}"
        super().__init__(name)
        self.my_process = process 
        self.my_location = location
        self.resource_executer = None
        self.next_location = None
        self.current_skill_index = 0

    def Setup(self, emaInteraction):
        self.AddBehavior(NextSkillBehaviour(self))

    def currentSkill(self):
        if self.current_skill_index < len(self.my_process):
            return self.my_process[self.current_skill_index]
        return None

    def hasNextSkill(self):
        return self.current_skill_index + 1 < len(self.my_process)

    def setNextSkill(self):
        self.current_skill_index += 1

class NextSkillBehaviour(OneShotBehavior):
    def Action(self):
        skill = self.agent.currentSkill()
        if skill is None:
            print(f"[{self.agent.getName()}] All skills completed.")
            return

        print(f"[{self.agent.getName()}] Searching for skill: {skill}")
        
        searchAD = AgentDescription()
        sd = ServiceDescription(name=skill)
        searchAD.AddService(sd)
        
        result = self.emaInteraction.Search(searchAD)
        
        if not result:
            print(f"[{self.agent.getName()}] No agents found for skill {skill}. Retrying...")
            time.sleep(1)
            self.agent.AddBehavior(NextSkillBehaviour(self.agent))
            return

        cfp = ACLMessage(Performative.CFP)
        for ad in result:
            cfp.receiver.append(ad.agentid)
        
        cfp.ontology = ONTOLOGY_NEGOTIATE_NEXT_RESOURCE
        cfp.content = skill
        
        self.agent.AddBehavior(CNETProposalInitiator(self.agent, cfp))

class CNETProposalInitiator(FIPAContractNetInitiator):
    def handleProposals(self, proposals, responses):
        if not proposals:
            print(f"[{self.agent.getName()}] No proposals received. Retrying NextSkillBehaviour...")
            self.agent.AddBehavior(NextSkillBehaviour(self.agent))
            return False

        # Pick the first proposal
        best_proposal = proposals[0]
        accept = best_proposal.CreateReply(Performative.ACCEPT_PROPOSAL)
        responses.append(accept)
        
        # Reject others
        for i in range(1, len(proposals)):
            reject = proposals[i].CreateReply(Performative.REJECT_PROPOSAL)
            responses.append(reject)
        
        return True

    def handleInform(self, msg):
        self.agent.resource_executer = msg.sender
        self.agent.next_location = msg.content
        
        print(f"[{self.agent.getName()}] Proposal accepted by {msg.sender.GetFullID()}. Next location: {self.agent.next_location}")
        
        # Search for transport
        searchAD = AgentDescription()
        sd = ServiceDescription(name=DF_SERVICE_NAME_TRANSPORT, type=DF_SERVICE_TYPE_TRANSPORT)
        searchAD.AddService(sd)
        
        result = self.emaInteraction.Search(searchAD)
        
        if not result:
            print(f"[{self.agent.getName()}] Transport agent not found!")
            return True

        req = ACLMessage(Performative.REQUEST)
        req.receiver.append(result[0].agentid)
        req.ontology = ONTOLOGY_REQUEST_TRANSPORT
        
        # Based on transport_module.py, use "b" for target location
        content = {
            "a": self.agent.my_location,
            "b": self.agent.next_location
        }
        req.content = json.dumps(content)
        
        self.agent.AddBehavior(REQUESTExecuteTransportInitiator(self.agent, req))
        return True

class REQUESTExecuteTransportInitiator(FIPARequestInitiator):
    def handleInform(self, msg):
        self.agent.my_location = self.agent.next_location
        print(f"[{self.agent.getName()}] Transport completed. Arrived at {self.agent.my_location}")
        
        req = ACLMessage(Performative.REQUEST)
        req.receiver.append(self.agent.resource_executer)
        req.ontology = ONTOLOGY_REQUEST_EXECUTE_SKILL
        req.content = self.agent.currentSkill()
        
        self.agent.AddBehavior(REQUESTExecuteSkillInitiator(self.agent, req))
        return True

    def handleRefuse(self, msg):
        print(f"[{self.agent.getName()}] Transport refused. Retrying NextSkillBehaviour...")
        self.agent.AddBehavior(NextSkillBehaviour(self.agent))
        return True

class REQUESTExecuteSkillInitiator(FIPARequestInitiator):
    def handleInform(self, msg):
        print(f"[{self.agent.getName()}] Skill {self.agent.currentSkill()} completed.")
        
        if self.agent.hasNextSkill():
            self.agent.setNextSkill()
            self.agent.AddBehavior(NextSkillBehaviour(self.agent))
        else:
            print(f"[{self.agent.getName()}] All skills completed. Returning to A.")
            
            # Transport back to A
            searchAD = AgentDescription()
            sd = ServiceDescription(name=DF_SERVICE_NAME_TRANSPORT, type=DF_SERVICE_TYPE_TRANSPORT)
            searchAD.AddService(sd)
            result = self.emaInteraction.Search(searchAD)
            
            if result:
                req = ACLMessage(Performative.REQUEST)
                req.receiver.append(result[0].agentid)
                req.ontology = ONTOLOGY_REQUEST_TRANSPORT
                content = {
                    "a": self.agent.my_location,
                    "b": "A" 
                }
                req.content = json.dumps(content)
                self.agent.AddBehavior(TransportBackBehaviour(self.agent, req))
            else:
                self.agent.Stop()
        return True

    def handleRefuse(self, msg):
        print(f"[{self.agent.getName()}] Skill execution refused. Retrying NextSkillBehaviour...")
        self.agent.AddBehavior(NextSkillBehaviour(self.agent))
        return True

class TransportBackBehaviour(FIPARequestInitiator):
    def handleInform(self, msg):
        print(f"[{self.agent.getName()}] Back at A. Stopping.")
        self.agent.Stop()
        return True
    
    def handleFailure(self, msg):
        print(f"[{self.agent.getName()}] Failed to return to A. Stopping.")
        self.agent.Stop()
        return True
